dhall/src/Dhall/Repl.hs

```diff
@@ -255,26 +254,138 @@ hashBinding tokens = do
 
   normalizedExpression <- normalize loadedExpression
 
-  Env{_standardVersion} <- get
+  Env{_standardVersion, outputHandle} <- get
 
-  liftIO . Text.putStrLn $
+  liftIO . System.IO.hPutStrLn outputHandle . Text.unpack $
     hashExpressionToCode _standardVersion normalizedExpression
 
+saveFilePrefix :: FilePath
+saveFilePrefix = ".dhall-repl"
+
+-- | Find the index for the current _active_ dhall save file
+currentSaveFileIndex :: MonadIO m => m (Maybe Int)
+currentSaveFileIndex = do
+  files <- liftIO $ getDirectoryContents "."
+
+  let parseIndex file
+        | saveFilePrefix `isPrefixOf` file
+        , '-':index <- drop (length saveFilePrefix) file
+        , [(x, "")] <- reads index -- safe version of read
+        = Just x
+
+        | otherwise
+        = Nothing
+
+  pure $ case mapMaybe parseIndex files of
+    [] -> Nothing
+    xs -> Just $ maximum xs
+
+-- | Find the name for the current _active_ dhall save file
+currentSaveFile :: MonadIO m => m (Maybe FilePath)
+currentSaveFile =
+  (fmap . fmap) (\i -> saveFilePrefix <> "-" <> show i) currentSaveFileIndex
+
+-- | Find the name for the next dhall save file
+nextSaveFile :: MonadIO m => m FilePath
+nextSaveFile = do
+  mIndex <- currentSaveFileIndex
+
+  let nextIndex = maybe 0 succ mIndex
+
+  pure $ saveFilePrefix <> "-" <> show nextIndex
+
+loadBinding
+  :: ( MonadIO m, MonadState Env m, Haskeline.MonadException m )
+  => [String] -> m ()
+loadBinding [] = do
+  mFile <- currentSaveFile
+
+  case mFile of
+    Just file -> loadBinding [file]
+    Nothing   ->
+      fail $ ":load couldn't find any `" <> saveFilePrefix <> "-*` files"
+
+loadBinding [file] = do
+  -- Read commands from the save file
+  replLines <- map words . lines <$> liftIO (readFile file)
+
+  -- Keep current handle in scope
+  env@Env { outputHandle } <- get
+
+  let runCommand ((c:cmd):opts)
+        | c == optionsPrefix
+        , Just action <- lookup cmd options
+        = action opts
+      runCommand _ = fail $
+        ":load expects `" <> file <> "` to contain one command per line"
+
+      -- Run all the commands
+      handler nullHandle =
+          State.execStateT
+            (forM_ replLines runCommand)
+            (env { outputHandle = nullHandle })
+
+  -- Use a null handle (to discard output) to run all the commands
+  newEnv <- liftIO $ System.IO.withFile "/dev/null" System.IO.WriteMode handler
```

Comment:
I have no idea how platform independent this solution is. Any pointers to improve this? Or do you think this works on Windows somehow? :crossed_fingers:

Reply:
Appveyor is happy so it works in windows 👍

Reply:
Unfortunately I don't think Appveyor runs the REPL, and this will likely be a runtime error 😅

Reply:
oh right, i sometimes forgot that sometimes it compiles but fails at runtime. It seems that it will fail in windows, yes.

Reply:
I've tested the pr in my windows env and i've not been able to trigger any error. The steps were:
* run the dhall repl
* type some binding: `:let a = 1`
* `:save` the session
* `:load` the session
* quit and rerun repl and load the session again

Reply:
Thanks for testing, I'm still surprised it doesn't crash. Could you try to exit in between `save` and `load`? And the check if `a` is in scope.

Reply:
Because of the `withFile` pattern, if there's an error in the handle. It probably just skips the loading bit

Reply:
> Thanks for testing, I'm still surprised it doesn't crash. Could you try to exit in between `save` and `load`? And the check if `a` is in scope.

My last test was just that, including the console output to be more precise:
```console
D:\dhall-haskell>stack run dhall -- repl

⊢ :let a = 1

a : Natural

⊢ a

1

⊢ :save

Context saved to `.dhall-repl-0`

⊢ :quit
Goodbye.

D:\dhall-haskell>stack run dhall -- repl

⊢ :load

Loaded `.dhall-repl-0`

⊢ a

1

Reply:
I am on windows 10, not sure if it would work in older systems.
Maybe the solution of this so answer could be applicable here: https://stackoverflow.com/a/41527205/49554

Reply:
Ah thank you! If `hSilence` is properly cross platform I could use that. Otherwise I would use a `Maybe Handle` as part of the state instead and just not print anything when I don't have a handle.
That might be cleaner! 😄

Reply:
Yeah, I would configure the REPL so it can optionally not print output when replaying commands